src/metatensor/models/utils/loss.py

```diff
@@ -0,0 +1,118 @@
+from typing import Dict, Optional
+
+import torch
+from metatensor.torch import TensorMap
+
+
+# This file defines losses for metatensor models.
+
+
+class TensorMapLoss:
+    """
+    A loss function that operates on two `metatensor.torch.TensorMap`s.
+
+    The loss is computed as the sum of the loss on the block values and
+    the loss on the gradients, with weights specified at initialization.
+
+    At the moment, this loss function assumes that all the gradients
+    declared at initialization are present in both TensorMaps.
+    """
+
+    def __init__(
+        self,
+        reduction: str = "mean",
+        weight: float = 1.0,
+        gradient_weights: Optional[Dict[str, float]] = None,
+    ):
+        self.loss = torch.nn.MSELoss(reduction=reduction)
+        self.weight = weight
+        self.gradient_weights = {} if gradient_weights is None else gradient_weights
+
+    def __call__(
+        self, tensor_map_1: TensorMap, tensor_map_2: TensorMap
+    ) -> torch.Tensor:
+        # Assert that the two have the same metadata, except for the samples,
+        # which can be different due to batching, but must have the same size:
+        assert tensor_map_1.keys == tensor_map_2.keys
+        assert tensor_map_1.block().properties == tensor_map_2.block().properties
+        assert tensor_map_1.block().components == tensor_map_2.block().components
+        assert len(tensor_map_1.block().samples) == len(tensor_map_2.block().samples)
+        for gradient_name in self.gradient_weights.keys():
+            assert len(tensor_map_1.block().gradient(gradient_name).samples) == len(
+                tensor_map_2.block().gradient(gradient_name).samples
+            )
+            assert (
+                tensor_map_1.block().gradient(gradient_name).properties
+                == tensor_map_2.block().gradient(gradient_name).properties
+            )
+            assert (
+                tensor_map_1.block().gradient(gradient_name).components
+                == tensor_map_2.block().gradient(gradient_name).components
+            )
```

Comment:
I don't know if these should really be asserts? This is user facing function and the user can give any tensor here and therefore proper errors i.e. `ValueErrors` with error messages should be raised.

Reply:
Well, the way we use it now, the user doesn't use it. But if we move it to metatensor-learn, then yes, it becomes user-facing. Since someone has to do it at some point, I'll turn them into ValueErrors